src/runner.rs

```diff
@@ -0,0 +1,124 @@
+use crate::coordinator::{self, RunnableJob};
+use anyhow::{Context, Result};
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+#[derive(Debug, Default)]
+pub struct Runner {
+    root: PathBuf,
+}
+
+impl Runner {
+    pub fn new(root: PathBuf) -> Self {
+        Runner { root }
+    }
+}
+
+impl coordinator::Runner for Runner {
+    fn run(&self, job: &RunnableJob) -> Result<()> {
+        let workspace = Workspace::create(&self.root, job)?;
+
+        debug_assert!(job.inputs.is_empty(), "we don't handle inputs yet");
+        debug_assert!(
+            job.input_files.is_empty(),
+            "we don't handle input files yet"
+        );
+
+        let mut command: Command = job.into();
+        command.current_dir(&workspace);
+
+        // TODO: send stdout, stderr, etc to The Log Zone(tm)
+        // TODO: rearrange this so we can stream logs
+        let status = command
+            .spawn()
+            .context("could not run command")?
+            .wait()
+            .context("command wasn't running")?;
+
+        match status.code() {
+            Some(0) => (),
+            Some(code) => anyhow::bail!("command failed with the exit code {code}"),
+            None => anyhow::bail!("command failed with no exit code (maybe it was killed?)"),
+        }
+
+        let store = Store::create(&self.root, job)?;
+        store.take_outputs_from_workspace(job, &workspace)?;
+
+        Ok(())
+    }
+}
+
+struct Workspace(PathBuf);
+
+impl Workspace {
+    fn create(root: &Path, job: &RunnableJob) -> Result<Self> {
+        let workspace = Workspace(root.join("workspaces").join(job.id.to_string()));
+
+        std::fs::create_dir_all(&workspace.0).context("could not create workspace")?;
```

Comment:
Shouldn't this work with the AsRef<Path> defined below instead of needing .0?

Reply:
oh, so it does! How nice! Still getting used to some of these traits.